### PR TITLE
ESLint Workflow

### LINKED ISSUE #29: Github Workflow
Setup a github workflow file to run tests and check linting

### PR DESCRIPTION
**Pull request by tbmcallister**
This PR closes #29

This PR just adds in the ESLint workflow, found in the GitHub actions page. I created the workflow and added it into the correct file path

./.github/workflows/eslint.yml

This will just run ESLint, which should help keep our code much cleaner overall. If something is left in the PR that shouldn't be there(unused code, etc.), this should help combat that issue, and we should have much simpler PR reviews as a result.

Future improvements will need to be made, such as creating a new workflow that will run our unit tests, after those are created.
I will be creating a new issue to encapsulate this improvement, after this PR is completed.

Please let me know if this needs improvement, and thank you!

### CODE REVIEW

**Comment by Linkaloo** [comment]
Some things to keep in mind while doing this:
- Prettier is part of the eslint rules, specifically in charge of code formatting
- The server and client folders have their own eslint rules
- To make performance for the workflow a little better try to see if you can specifically target eslint for one of the server or client files based on which directory has changes, idk if that is possible but worth looking into

**Comment by tbmcallister** [comment]
the server one is working, I just need to figure out the client side issue(the error message im getting is not at all descriptive).

**Comment by KingJordan152** [comment]
Maybe this should be labelled as a draft... 👀

**Comment by tbmcallister** [comment]
alright. i think i am crazy, but i think github is more crazy. I believe my PR can be reviewed for the ESLint Workflow. I believe that the client one has findings, which is why it is returning an "Error: Process completed with exit code 1.". I'm not sure why it's returning an error, but if someone has a suggestion I can implement it rq. But, the workflow is working im pretty sure. please ignore the commits, since its just me changing one value back and forth. the only thing you need to review is just the addition of the .github/workflows/eslint.yml

TLDR: workflow PR is ready for review :))

**Review by KingJordan152 — APPROVED** [review]
Everything looks good to me here. Given that the checks are passing on GitHub, it seems to me that this workflow is usable. I would suggest adding Stylelint checks and running Vite/Jest, but maybe that can be incorporated in a future PR. I appreciate your hard work and dedication; I know this wasn't easy. Great work!

**Review by morenj — APPROVED** [review]
Amazing work here Tim! Everything looks good for this PR and it seems you fixed the tests passing. Great work overall, speedy results for a challenging task!

I agree with saving the additional implementations for a future PR. Thank you again for another well done PR.

File: .eslintrc.js
Comment (Linkaloo) [inline .eslintrc.js:1]: I am going to ask you to go back in the trenches and try to remove this file.
Couple of reasons:
I am worried that this can interfere with our linting rules on the server and client folders.
It should not be needed because the workflow file goes into the client and server folder which have eslint rules defined.

File: package.json [hunk 1]
```diff
@@ -19,5 +19,11 @@
     "@mui/icons-material": "^5.15.13",
     "@mui/material": "^5.15.13",
     "js-cookie": "^3.0.5"
+  },
+  "devDependencies": {
+    "@typescript-eslint/eslint-plugin": "^7.3.1",
+    "@typescript-eslint/parser": "^7.3.1",
+    "eslint": "^8.56.0",
+    "eslint-plugin-react": "^7.34.1"
```
Comment (Linkaloo) [inline package.json:27]: Since we should remove the eslint file from the root we should also remove these 2 eslint dependencies.
npm uninstall eslint eslint-plugin-react.

File: .github/workflows/eslint.yml [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      - name: Checkout code
+        uses: actions/checkout@v3
+
+      - name: Install ESLint
+        working-directory: ./client
+        run: |
+          npm install eslint@8.56.0
+          npm install @microsoft/eslint-formatter-sarif@2.1.7
+          npm install eslint-plugin-react@latest --save-dev
+          npm install @typescript-eslint/eslint-plugin@latest --save-dev --force
```
Comment (Linkaloo) [inline .github/workflows/eslint.yml:39]: Try to just do npm install.
You're inside the client directory already which has a package file. You should just be able to do an npm install and then run the eslint command. It's fine if other packages are installed.
If you continue to do this then you'd have to manage this workflow file for any change to eslint and that sounds not fun.

File: .github/workflows/eslint.yml [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
+
+      - name: Install ESLint
+        working-directory: ./server
+        run: |
+          npm install eslint@8.56.0
+          npm install @microsoft/eslint-formatter-sarif@2.1.7
+          npm install eslint-plugin-react@latest --save-dev
```
Comment (Linkaloo) [inline .github/workflows/eslint.yml:72]: Same here try npm install and then run eslint

File: .github/workflows/eslint.yml [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+          npm install @typescript-eslint/eslint-plugin@latest --save-dev --force
+
+      - name: Run ESLint
+        working-directory: ./client
+        run: npx eslint .
+          --config .eslintrc.cjs
+          --ext .js,.jsx,.ts,.tsx
+          --format @microsoft/eslint-formatter-sarif
+          --output-file eslint-results.sarif
+        continue-on-error: true
```
Comment (Linkaloo) [inline .github/workflows/eslint.yml:48]: I understand you got angry about this and probably just let it continue, but that defeats the point of the workflow file, so this should be removed.

File: .github/workflows/eslint.yml [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+          npm install eslint-plugin-react@latest --save-dev
+
+      - name: Run ESLint
+        working-directory: ./server
+        run: npx eslint .
+          --config .eslintrc.json
+          --ext .js,.jsx,.ts,.tsx
+          --format @microsoft/eslint-formatter-sarif
+          --output-file eslint-results.sarif
+        continue-on-error: true
```
Comment (Linkaloo) [inline .github/workflows/eslint.yml:81]: Same with this continue-on-error, it should be removed

**Review by Linkaloo — CHANGES_REQUESTED** [review]
I'm sorry but I feel like there are a lot of things that should change in here.
Since this is a workflow pr I am being harsher with it cause its meant to help us

File: package.json (hunk 1 above)
Comment (tbmcallister) [reply package.json]: that was fixed in another commit, and i also did the npm uninstall. fixed

File: .github/workflows/eslint.yml (hunk 2 above)
Comment (tbmcallister) [reply .github/workflows/eslint.yml:39]: fixed, i was using a default template and this way makes a lot more sense

File: .github/workflows/eslint.yml (hunk 3 above)
Comment (tbmcallister) [reply .github/workflows/eslint.yml:72]: fixed

File: .github/workflows/eslint.yml (hunk 4 above)
Comment (tbmcallister) [reply .github/workflows/eslint.yml]: removed

File: .github/workflows/eslint.yml (hunk 5 above)
Comment (tbmcallister) [reply .github/workflows/eslint.yml]: removed

File: .eslintrc.js
Comment (tbmcallister) [reply .eslintrc.js:1]: this file was removed. it wasn't doing anything because of the "root":true in the other config files(since if the config has that, it will not search up the parent directory to find a config). BUT, this is better to have gone to be safe than sorry.

**Comment by tbmcallister** [comment]
the continue_on_error flag is important because if there is some sort of error, it will be uploaded to github so we can actually see what the error is(as far as i know and from what ive seen). i implemented your changes, but i keep getting a failed test, and no useful outputs. it also isn't uploading to github. i understand why you suggested it so that it fails when theres a finding, im just not sure how to implement this correctly.  i can work on this in a future PR, and also when we incorporate the unit testing into the workflow.

but for now, the workflow just points you in the right direction as to where the findings are, and the reviewers can verify that the code committed is not creating any findings.

**Comment by tbmcallister** [comment]
I will be creating a new issue to fix the findings found by this test run after this PR merges with main

**Review by Linkaloo — APPROVED** [review]
